rpm/bdb/bdb.go

```diff
@@ -0,0 +1,297 @@
+	Metaflags   byte     /* 26: Meta-only flags */
+	_           byte     /* 27: Unused. */
+	Free        uint32   /* 28-31: Free list page number. */
+	LastPageNo  uint32   /* 32-35: Page number of last page in db. */
+	NParts      uint32   /* 36-39: Number of partitions. */
+	KeyCount    uint32   /* 40-43: Cached key count. */
+	RecordCount uint32   /* 44-47: Cached record count. */
+	Flags       uint32   /* 48-51: Flags: unique to each AM. */
+	UID         [20]byte /* 52-71: Unique file ID. */
+}
+
+// Pagetype numbers:
+const (
+	pagetypeHashMeta     = 8
+	pagetypeHashUnsorted = 2
+	pagetypeHash         = 13
+	pagetypeHashOffIndex = 3
+	pagetypeOverflow     = 7
+)
+
+// Serialized sizes:
+const (
+	hashpageSize    = 26
+	hashoffpageSize = 12
+)
+
+// Hash database metadata, aka HMETA in C.
+type hashmeta struct {
+	meta                   /* 00-71: Generic meta-data page header. */
+	MaxBucket   uint32     /* 72-75: ID of Maximum bucket in use */
+	HighMask    uint32     /* 76-79: Modulo mask into table */
+	LowMask     uint32     /* 80-83: Modulo mask into table lower half */
+	FllFactor   uint32     /* 84-87: Fill factor */
+	NElem       uint32     /* 88-91: Number of keys in hash table */
+	HashCharKey uint32     /* 92-95: Value of hash(CHARKEY) */
+	_           [32]uint32 /* 96-223: Spare pages for overflow */
+	_           [59]uint32 /* 224-459: Unused space */
+	CryptoMagic uint32     /* 460-463: Crypto magic number */
+	_           [3]uint32  /* 464-475: Trash space - Do not use */
+	// The comments don't line up, but the numbers come from the source, so...
+	IV       [16]byte /* 476-495: Crypto IV */
+	Checksum [20]byte /* 496-511: Page chksum */
+}
+
+// Hash page header, aka PAGE in C.
+//
+// Also shared with btree databases, which are unimplemented here.
+// The [meta.PageSize] block of memory has this struct at position 0, then
+// populates it backwards from the end for structured data, or immediately after
+// this for binary data.
+type hashpage struct {
+	LSN            [8]byte /* 00-07: Log sequence number. */
+	PageNo         uint32  /* 08-11: Current page number. */
+	PrevPageNo     uint32  /* 12-15: Previous page number. */
+	NextPageNo     uint32  /* 16-19: Next page number. */
+	Entries        uint16  /* 20-21: Number of items on the page. */
+	HighFreeOffset uint16  /* 22-23: High free byte page offset. */
+	Level          byte    /*    24: Btree tree level. */
+	Type           byte    /*    25: Page type. */
+}
+
+// Hash offpage header, aka HOFFPAGE in C.
+//
+// This stores the data on how to extract "overflow"/"offpage" data.
+type hashoffpage struct {
+	Type   byte    /*    00: Page type and delete flag. */
+	_      [3]byte /* 01-03: Padding, unused. */
+	PageNo uint32  /* 04-07: Offpage page number. */
+	Length uint32  /* 08-11: Total length of item. */
+}
+
+// AllHeaders returns ReaderAts for all RPM headers in the PackageDB.
+func (db *PackageDB) AllHeaders(ctx context.Context) ([]io.ReaderAt, error) {
```

Comment:
nit: it might be nice to prefix these errors with `bdb:` for consistency and debugging ease